auctions/english-auction/src/data_structures.sw

```diff
@@ -0,0 +1,170 @@
+library data_structures;
+
+dep errors;
+
+use errors::AssetError;
+
+pub enum Asset {
+    NFTAsset: NFTAsset,
+    TokenAsset: TokenAsset,
+}
+
+impl Asset {
+    pub fn amount(self) -> u64 {
+        match self {
+            Asset::NFTAsset(nft_asset) => {
+                // TODO: This should return the number of tokens when StorageVec in structs is supported
+                1
+            },
+            Asset::TokenAsset(token_asset) => {
+                token_asset.amount
+            },
+        }
+    }
+
+    pub fn contract_id(self) -> ContractId {
+        match self {
+            Asset::NFTAsset(nft_asset) => {
+                nft_asset.contract_id
+            },
+            Asset::TokenAsset(token_asset) => {
+                token_asset.contract_id
+            },
+        }
+    }
+}
+
+impl core::ops::Add for Asset {
+    fn add(self, other: Self) -> Self {
+        match (self, other) {
+            (
+
+                Asset::TokenAsset(token_asset1),
+                Asset::TokenAsset(token_asset2),
+            ) => {
+                require(token_asset1.contract_id == token_asset2.contract_id, AssetError::AssetsAreNotTheSame);
+                let token = TokenAsset {
+                    amount: token_asset1.amount + token_asset2.amount,
+                    contract_id: token_asset1.contract_id,
+                };
+                Asset::TokenAsset(token)
```

Comment:
I'm not sure what you're referring to here

Reply:
Instead of assigning the struct to a variable and passing that variable into the enum the struct instantiation can be done directly inside the enum which alleviates the need to allocate space for a variable.